File: roomsplit/frontend/src/types/payment.ts
export type PaymentStatus = 'pending' | 'marked_paid' | 'confirmed' | 'rejected' | 'expired'
export type PaymentType = 'rent' | 'expense' | 'withdrawal'

export interface Payment {
  id: string
  group_id: string
  cycle_id?: string
  payer_id: string
  payer_name: string
  amount_paise: number
  amount_rupees: number
  upi_ref?: string
  payment_type: PaymentType
  status: PaymentStatus
  due_date?: string
  marked_at?: string
  confirmed_at?: string
  rejection_reason?: string
  created_at: string
}

export interface UPILinkResponse {
  upi_link: string
  qr_code_base64: string
  amount_rupees: number
  payee_name: string
  payee_upi: string
}

export interface RentCycle {
  id: string
  group_id: string
  label: string
  period_start: string
  period_end: string
  total_amount_paise: number
  total_amount_rupees: number
  due_date: string
  status: 'draft' | 'active' | 'closed' | 'cancelled'
  created_at: string
}

export interface RentAssignment {
  id: string
  cycle_id: string
  member_id: string
  member_name: string
  assigned_amount_paise: number
  assigned_amount_rupees: number
  split_type: string
}

File: roomsplit/backend/app/models/payment.py
import uuid
from datetime import datetime, date
from typing import Optional
from sqlalchemy import BigInteger, Date, DateTime, ForeignKey, Enum, Text, String, func
from sqlalchemy.dialects.postgresql import UUID as PG_UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship
from app.database import Base
from app.models.base import TimestampMixin


class Payment(TimestampMixin, Base):
    __tablename__ = "payments"

    group_id: Mapped[uuid.UUID] = mapped_column(
        PG_UUID(as_uuid=True), ForeignKey("groups.id"), nullable=False, index=True
    )
    cycle_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        PG_UUID(as_uuid=True), ForeignKey("rent_cycles.id"), nullable=True, index=True
    )
    assignment_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        PG_UUID(as_uuid=True), ForeignKey("rent_assignments.id"), nullable=True
    )
    payer_id: Mapped[uuid.UUID] = mapped_column(
        PG_UUID(as_uuid=True), ForeignKey("users.id"), nullable=False, index=True
    )
    payee_id: Mapped[uuid.UUID] = mapped_column(
        PG_UUID(as_uuid=True), ForeignKey("users.id"), nullable=False
    )
    amount: Mapped[int] = mapped_column(BigInteger, nullable=False)  # paise
    upi_ref: Mapped[Optional[str]] = mapped_column(String(100), nullable=True)
    payment_type: Mapped[str] = mapped_column(
        Enum("rent", "expense", "withdrawal", name="payment_type_enum"),
        default="rent",
        nullable=False,
    )
    status: Mapped[str] = mapped_column(
        Enum(
            "pending",
            "marked_paid",
            "confirmed",
            "rejected",
            "expired",
            name="payment_status_enum",
        ),
        default="pending",
        nullable=False,
        index=True,
    )
    marked_at: Mapped[Optional[datetime]] = mapped_column(DateTime(timezone=True), nullable=True)
    confirmed_at: Mapped[Optional[datetime]] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    confirmed_by: Mapped[Optional[uuid.UUID]] = mapped_column(
        PG_UUID(as_uuid=True), ForeignKey("users.id"), nullable=True
    )
    rejected_at: Mapped[Optional[datetime]] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    rejected_by: Mapped[Optional[uuid.UUID]] = mapped_column(
        PG_UUID(as_uuid=True), ForeignKey("users.id"), nullable=True
    )
    rejection_reason: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    due_date: Mapped[Optional[date]] = mapped_column(Date, nullable=True)

    # Relationships
    group: Mapped["Group"] = relationship("Group")  # noqa: F821
    cycle: Mapped[Optional["RentCycle"]] = relationship(  # noqa: F821
        "RentCycle", back_populates="payments"
    )
    assignment: Mapped[Optional["RentAssignment"]] = relationship("RentAssignment")  # noqa: F821
    payer: Mapped["User"] = relationship("User", foreign_keys=[payer_id])  # noqa: F821
    payee: Mapped["User"] = relationship("User", foreign_keys=[payee_id])  # noqa: F821
    confirmer: Mapped[Optional["User"]] = relationship(  # noqa: F821
        "User", foreign_keys=[confirmed_by]
    )
    rejecter: Mapped[Optional["User"]] = relationship(  # noqa: F821
        "User", foreign_keys=[rejected_by]
    )
